File: 496.py
# Using monotonic stack
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        map = {}
        stack = []

        for i in range(len(nums2) - 1, -1, -1):
            num = nums2[i]

            while stack and stack[-1] <= num:
                stack.pop()
            
            map[num] = stack[-1] if stack else -1
            stack.append(num)
            
        return [map[num] for num in nums1]
        

# class Solution:
#     def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
#         map = {}
#         n = len(nums2)

#         for i in range(n):
#             key = nums2[i]
#             value = -1
#             for j in range(i + 1, n):
#                 if nums2[j] > key:
#                     value = nums2[j]
#                     break
            
#             map[key] = value

#         return [map[num] for num in nums1]

File: 496.java
class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        Map<Integer, Integer> map = new HashMap<>();

        int n = nums2.length;
        for(int i = 0; i < n; i++) {
            int key = nums2[i];
            int value = -1;
            for(int j = i + 1; j < n; j++) {
                if(nums2[j] > key) {
                    value = nums2[j];
                    break;
                }
            }   
            map.put(key, value);
        }

        int result[] = new int[nums1.length];
        for(int i = 0; i < nums1.length; i++) {
            result[i] = map.get(nums1[i]);
        }
        return result;
    }
}
